api/v1alpha1/openstackimage_types.go

```diff
@@ -0,0 +1,218 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// See https://docs.openstack.org/glance/latest/admin/useful-image-properties.html
+// for a list of 'well known' image properties we might consider supporting explicitly.
+
+// +kubebuilder:validation:MinLength:=1
+// +kubebuilder:validation:MaxLength:=255
+type ImageTag string
+
+// +kubebuilder:validation:MinLength:=1
+// +kubebuilder:validation:MaxLength:=10
+type ImageContainerFormat string
+
+// +kubebuilder:validation:MinLength:=1
+// +kubebuilder:validation:MaxLength:=10
+type ImageDiskFormat string
+
+// kubebuilder:validation:Enum:=public;private;shared;community
+type ImageVisibility string
+
+// kubebuilder:validation:Enum:=md5;sha1;sha256;sha512
+type ImageHashAlgorithm string
+
+const (
+	ImageHashAlgorithmMD5    ImageHashAlgorithm = "md5"
+	ImageHashAlgorithmSHA1   ImageHashAlgorithm = "sha1"
+	ImageHashAlgorithmSHA256 ImageHashAlgorithm = "sha256"
+	ImageHashAlgorithmSHA512 ImageHashAlgorithm = "sha512"
+)
+
+// +kubebuilder:validation:XValidation:rule="!in(self, ['created_at', 'updated_at', 'status', 'checksum', 'size', 'virtual_size', 'direct_url', 'self', 'file', 'schema', 'id', 'os_hash_algo', 'os_hash_value', 'location', 'deleted', 'deleted_at', 'container_format', 'disk_format', 'min_disk', 'min_ram', 'name', 'tags', 'owner', 'visibility', 'protected', 'os_hidden'])"
+type ImageAdditionalPropertyName string
+
+type ImageAdditionalProperty struct {
+	// Name is the name of the glance property. It is an error for it to conflict with any explicitly defined property.
+	Name ImageAdditionalPropertyName `json:"name"`
+
+	// Value is the value of the glance property
+	// +required
+	// +kubebuilder:validation:MinLength:=1
+	// +kubebuilder:validation:MaxLength:=1024
+	Value string `json:"value"`
+}
+
+type ImageProperties struct {
+	// MinDisk is the minimum amount of disk space in GB that is required to boot the image
+	// +kubebuilder:validation:Minimum:=1
```

Comment:
What about a maximum value?

Reply:
Deliberately left it out here. I don't think it has a sensible maximum value. It's basically just a (server-side) validation that the server you're booting has sufficient disk space for the intended workload. Putting an enormous value here is just going to prevent anybody from booting from it; there is no associated resource usage. It's also mutable, so they can fix it.